sysutils/autoswc/files/autoswc.8
.\" $NetBSD$
.\"
.\" autoswc - Generates system wide cache files for GNU autoconf
.\" Copyright (c) 2004 Julio M. Merino Vidal <jmmv@NetBSD.org>
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Neither the name of The NetBSD Foundation nor the names of its
.\"    contributors may be used to endorse or promote products derived
.\"    from this software without specific prior written permission.
.\" 3. Neither the name of author nor the names of its contributors may
.\"    be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd February 21, 2004
.Dt AUTOSWC 8
.Os
.Sh NAME
.Nm autoswc
.Nd generates system wide cache files for GNU autoconf
.Sh SYNOPSIS
.Nm
.Op Fl c Ar ac_file
.Op Fl n Ar cache_name
.Sh DESCRIPTION
.Nm
is an utility that automatically generates system wide cache files for use
with configure scripts created by
.Nm GNU autoconf ,
and specially by those executed within
.Nm pkgsrc .
.Pp
.Nm
starts by reading the configuration script template.
It will use
.Pa @SYSCONFDIR@/configure.ac
if found, or fallback to
.Pa @PREFIX@/share/autoswc/configure.ac
otherwise (a custom file may be given using the
.Fl c
flag, though).
This issues all checks that the user wants to store in the system wide
cache file, and are basically calls to standard
.Nm GNU autoconf ,
.Nm GNU automake
and/or
.Nm GNU libtool
macros.
There is no need to manually execute any of the
.Sq AC_INIT ,
.Sq AM_INIT_AUTOMAKE ,
.Sq AC_CACHE_SAVE
and
.Sq AC_OUTPUT
macros, as they are automatically inserted in the right place during
execution.
.Pp
The default cache file is stored inside the
.Pa @CACHEDIR@
directory, using
.Pa config.cache.default
as the file name.
The
.Sq default
part can be changed using the
.Fl n
flag, which takes the base name of the cache file.
.Pp
To tell
.Nm pkgsrc
to use the system wide cache, all you need to do is append the following
line to your
.Pa @MKCONF@
file:
.Bd -literal -offset indent
\&.sinclude "@PREFIX@/share/autoswc/autoswc.mk"
.Ed
.Pp
In case you used the
.Fl n
flag, you also need to define the
.Va AUTOSWC_CACHE_NAME
variable in this file, setting it to the value passed to the flag.
.Sh NOTES
Be very careful about what you add to
.Pa configure.ac .
You should only check for things that are unlikely to change over time,
as functions in the
.Lb libc
and its header files.
Failure to do so can easily lead to cache files that get outdated very
easily during installation of new packages, and cause unexpected
configuration errors.
.Pp
It is recommended that you rebuild the cache file periodically.
You can configure a simple
.Xr cron 8
job to it, which just needs to launch this utility.
.Pp
If you hit a compilation problem in any package when using
.Nm ,
be
.Em absolutely
sure it is not beeing caused by an outdated cache before reporting
problems.
The safest and easiest way to do this is to clean the package in question
and restart the build with the
.Va AUTOSWC_DISABLE
variable set, like in the following example:
.Bd -literal -offset indent
$ @MAKE@ clean
$ @MAKE@ AUTOSWC_DISABLE=YES
.Ed
.Sh AUTHORS
.An Julio M. Merino Vidal Aq jmmv@netbsd.org


doc/NEWS
# $NetBSD: NEWS,v 1.1 2008/01/18 05:47:30 tnn Exp $

PKGSRC NEWS
===========

  This file contains information about important infrastructural changes in
  pkgsrc. The intended readers are the pkgsrc developers as well as anyone
  tracking pkgsrc-current.


OpenSSL update [tnn 20080118]
	OpenSSL in pkgsrc switched from the 0.9.7 to the 0.9.8 branch. To
	accomodate for this, 596 packages dependent on openssl had their
	revisions bumped.

pkgsrc-2007Q4 released [agc 20080105]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2008/01/05/0000.html

Simplifying meta-packages [rillig 20071219]
	The variable META_PACKAGE was introduced. This simplifies creation
	and maintenance of packages that do not install files.

Replace a package with different name [gdt 20071107]
	The variable OLDNAME was introduced for "make replace" to
	replace a package with a name other than the new one, via
	"make replace OLDNAME=oldpkgname".  Read the announcement at:
	http://mail-index.netbsd.org/tech-pkg/2007/11/07/0002.html
	
pkgsrc-2007Q3 released [agc 20071015]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/10/15/0000.html

Features framework [jlam 20070907]
	A "system features" framework has been added. It will be used to
	automatically supply missing "basic" headers and libraries for old
	systems. This is supported through the USE_FEATURES variable.

X.org 6.9 removed [joerg 20070817]
	monolithic X.org 6.9 was removed from pkgsrc. It was obsoleted by the
	modular X.org (7.x) packages. Transitioning users should remove all old
	X.org packages and set X11_TYPE=modular in mk.conf.

Emulator framework [jlam 20070729]
	A new emulator framework has been added. This will handle all
	binary-only packages that require emulation on the native OS.
	Please see pkgsrc/mk/emulator/README.

audit-packages merged with pkg_install [adrianp 20070714]
	The utility for auditing installed packages for security problems,
	audit-packages, has been rewritten in C. It is now part of the
	pkgtools/pkg_install package.
	
pkgsrc-2007Q2 released [agc 20070701]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/07/01/0000.html

New framework for bulk builds [joerg 20070619]
	pbulk, the new automated pkgsrc bulk build framework was imported.

pkgsrc-2007Q1 released [agc 20070419]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/04/19/0000.html

pkgsrc-2006Q4 released [agc 20070104]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/01/04/0000.html

modular X.org imported [joerg 20061114]
	The modular X.org packages were imported.

pkgsrc-2006Q3 released [agc 20061001]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2006/10/01/0000.html


pkgtools/autoswc/files/autoswc.sh
#!@SH@
#
# $NetBSD: autoswc.sh,v 1.1.1.1 2004/02/21 23:28:36 jmmv Exp $
#
# autoswc - Generates system wide cache files for GNU autoconf
# Copyright (c) 2004 Julio M. Merino Vidal <jmmv@NetBSD.org>
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in
#    the documentation and/or other materials provided with the
#    distribution.
# 3. Neither the name of author nor the names of its contributors may
#    be used to endorse or promote products derived from this software
#    without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
# ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
# TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
# BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#

ProgName=${0##*/}

err() {
    echo "${ProgName}: $*" 1>&2
    exit 1
}

info() {
    echo "${ProgName}: $*" 1>&2
}

mktempdir() {
    local d tmpdir

    # Check if mktemp exists in the current system.
    for d in $(echo ${PATH} | tr ':' ' '); do
        if [ -x $d/mktemp ]; then
            $d/mktemp -d /tmp/autoswc.XXXX
            return
        fi
    done

    # Emulate mktemp; this is not safe.
    info "mktemp(1) not found; race conditions may appear"
    tmpdir=/tmp/autoswc.$$
    mkdir -p ${tmpdir}
    chmod 700 ${tmpdir}
    echo ${tmpdir}
}

usage() {
    echo "Usage: ${ProgName} [-c ac_file] [-n cache_name]" 1>&2
    exit 1
}

# -------------------------------------------------------------------------

# Main program

cachename="default"

confsrc="@PREFIX@/share/autoswc/configure.ac"
[ -f "@SYSCONFDIR@/configure.ac" ] && confsrc="@SYSCONFDIR@/configure.ac"

# Parse command line options
args=$(getopt c:n: $*)
[ $? -ne 0 ] && usage
set -- ${args}
while [ $# -gt 0 ]; do
    case "$1" in
        -c)
            confsrc="$2"; shift
            ;;
        -n)
            cachename="$2"; shift
            ;;
        --)
            shift; break
            ;;
    esac
    shift
done
[ $# -gt 0 ] && usage

# Set the cache name.
confcache="@CACHEDIR@/config.cache.${cachename}"

# Check if we will be able to store the resulting cache file.
# The permissions may change during the execution of this script, but this
# won't bring us problems.  This check is just done to save some time in
# case we got the wrong user running autoswc.
touch ${confcache} >/dev/null 2>&1 || err "can't update ${confcache}"

# Ensure that the source configure.ac exists.
[ ! -f ${confsrc} ] && err "${confsrc} does not exist"

# Create work directory.
wrkdir=$(mktempdir)
info "using ${wrkdir} as work directory"
cd ${wrkdir}

# Create work project.  This includes:
# - Makefile.am - empty, just needs to be present.
# - configure.ac - the user-defined configure.ac file, modified to include
#                  basic and required macros.
# - NEWS README AUTHORS ChangeLog - empty, just need to be present.
echo "# empty" >Makefile.am

echo "AC_INIT(configure.ac)" >configure.ac
echo "AM_INIT_AUTOMAKE" >>configure.ac
cat ${confsrc} >>configure.ac
echo "AC_CACHE_SAVE" >>configure.ac
echo "AC_OUTPUT(Makefile)" >>configure.ac

touch NEWS README AUTHORS ChangeLog

# Generate scripts and add missing files.
info "running @ACLOCAL@..."
@ACLOCAL@
if grep LIBTOOL configure.ac >/dev/null 2>&1; then
    info "running @LIBTOOLIZE@..."
    @LIBTOOLIZE@ --force
fi
info "running @AUTOMAKE@..."
@AUTOMAKE@ --add-missing --force-missing
info "running @AUTOCONF@..."
@AUTOCONF@

# Run the configure script.
[ ! -f configure ] && err "autoconf failed"
CONFIG_SITE= ./configure --cache-file=config.cache \
    --build=@MACHINE_GNU_PLATFORM@ --host=@MACHINE_GNU_PLATFORM@ || \
    err "configure failed"
mv config.cache config.cache.in

# Prune the cache file from entries that will abort configure scripts when
# running from within pkgsrc.  This is not 100% safe, because configure is
# right in warning from changed values, but is needed to get this working.
grep -v "^ac_cv_env" <config.cache.in >config.cache

# Update the cache file.  We don't give it write permissions since we don't
# want third-party configure scripts update it with unwanted results.
install -c -m 444 config.cache ${confcache} || err "can't update ${confcache}"

cd -

rm -rf ${wrkdir}

# Start of summary.
cat <<EOF

===========================================================================
AUTOSWC - SUMMARY FOR \`${cachename}' CACHE

    ${confcache} updated successfully.

EOF

# Warn the user if autoswc.mk is not included from mk.conf.
if ! egrep '^.-?[ \t]*s?include[ \t]+"@PREFIX@/share/autoswc/autoswc.mk"' \
    @MKCONF@ >/dev/null 2>&1; then
    cat <<EOF
    The @MKCONF@ file seems to not include autoswc.mk.  This means that
    configure scripts will not find the generated cache file.  Add the
    following line to it:

    .sinclude "@PREFIX@/share/autoswc/autoswc.mk"

EOF
fi

# Tell the user to define AUTOSWC_CACHE_NAME if he used a custom cache
# name.
if [ ${cachename} != default ]; then
    cat <<EOF
    You have used a non-standard cache name.  To let configure scripts
    find it, you should add the following line somewhere in your
    @MKCONF@ file, in case it is not there yet:

    AUTOSWC_CACHE_NAME= ${cachename}

EOF
fi

# Warn the user about the dangers of using a system wide cache file.
cat <<EOF
    Please remember that by using a system wide cache file, you assume
    that the results stored in it may get outdated and cause strange build
    failures in third-party packages.  When reporting bugs, be absolutely
    sure it is not your fault.  Check the autoswc(8) manpage for more
    details.
===========================================================================
EOF

exit 0

# vim: syntax=sh:expandtab:shiftwidth=4:softtabstop=4
